pubtools/exodus/_hooks/pulp.py

```diff
@@ -0,0 +1,220 @@
+            total=int(self.retries),
+            backoff_factor=1,
+            status_forcelist=[429, 500, 502, 503, 504],
+        )
+        adapter = requests.adapters.HTTPAdapter(max_retries=retry_strategy)
+
+        out = requests.Session()
+        out.cert = (self.gw_crt, self.gw_key)
+        out.mount(self.gw_url, adapter)
+
+        return out
+
+    def unpack_response(self, response):
+        """Raise if response was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible.
+        """
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    def do_request(self, **kwargs):
+        if not self.session:
+            self.session = self.new_session()
+
+        resp = self.session.request(**kwargs)
+        self.unpack_response(resp)
+        return resp
+
+    def check_cert(self):
+        """Issue request to exodus-gw to identify permissions."""
+
+        for k, val in {"cert": self.gw_crt, "key": self.gw_key}.items():
+            if not val:
+                LOG.debug("exodus-gw %s not found, authentication may fail", k)
+
+        auth_url = urljoin(self.gw_url, "/whoami")
+        resp = self.do_request(method="GET", url=auth_url)
+        context = resp.json()
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                roles = [str(role) for role in typed_ctx["roles"]]
+                LOG.debug(
+                    "authenticated with exodus-gw at %s as %s %s (roles: %s)",
+                    self.gw_url,
+                    user_type,
+                    typed_ctx[ident],
+                    roles,
+                )
+                break
+        else:
+            raise RuntimeError(
+                "not authenticated with exodus-gw at %s" % self.gw_url
+            )
+
+    def new_publish(self):
+        """Issue request to exodus-gw to create a new publish."""
+        self.check_cert()
+
+        publish_url = os.path.join(self.gw_url, self.gw_env, "publish")
+        resp = self.do_request(method="POST", url=publish_url)
+        return resp.json()
+
+    def poll_commit_completion(self, commit):
+        """Issues request(s) to exodus-gw for the commit's state, returning
+        if/when the state is either "COMPLETE" or "FAILED".
+
+        Polls for a set amount of time (default 15 minutes) using a
+        fibonacci-based backoff.
+        """
+
+        def now():
+            return datetime.now(timezone.utc)
+
+        def fib(i):
+            return i if i <= 1 else fib(i - 1) + fib(i - 2)
```

Comment:
How about just dropping the idea of using fibonacci-based backoff or any exponential backoff?  Polling every ~15 seconds ought to be just fine.

As I understand it, when we use these kinds of "increase the delay on each poll" schemes, what we're trying to do is avoid the case where: the system hasn't finished a job yet, so more clients are waiting on their jobs and start polling it, so the processing of jobs slows down further, so yet more clients are waiting and start polling, etc...

I'm not sure it applies in cases like this where the thing processing a job is a separate process in a separate container from the http server, so polling does not affect the processing of jobs at all.

Also this is not a good implementation of fib as it will become very CPU-intensive and slow very quickly, for example paste this into a python session and see how long fib(40) takes on your machine.  Though it can be reasonably argued that the numbers won't get high enough here to matter.

Reply:
Yeah, I didn't anticipate this going over fib(10).

Okay. So, you suggest a hard-coded 15 sec wait. I think that may be too long. How about checking on the status every 5 seconds?